gamification/models.py

```diff
@@ -0,0 +1,202 @@
+    score = models.IntegerField(default=0, null=True)
+
+    # Current level
+    level = models.ForeignKey(Level, on_delete=models.CASCADE,
+        default=1, null=True)  # Ignore PycodestyleBear (E128)
+
+    # All the activities performed
+    activities = models.ManyToManyField(Activity)
+    badges = models.ManyToManyField(Badge)
+
+    def __str__(self):
+        return self.username
+
+    class Meta:
+        ordering = ['-score']
+
+    def add_points(self, points, activity_string):
+        """
+        Update score, level and add activities peformed.
+
+        :param points: an integer value representing the
+                       points earned by the newcomer
+                       for performing an activity
+        :param activity_string: represents the activity
+                                performed by the newcomer
+        """
+        self.update_score_and_level(points)
+        self.add_activity(points, activity_string)
+
+    def deduct_points(self, points_to_deduct, activity_string):
+        """
+        Deduct points for performing some specific activities.
+        """
+        self.add_points(-points_to_deduct, activity_string)
+
+    def find_level_for_score(self, score):
+        """
+        Find suitable level based on the total score earned.
+        """
+        level = Level.objects.get(min_score__lte=score, max_score__gt=score)
+        return level
+
+    def update_score_and_level(self, points):
+        """
+        Update score and level based on points.
+        """
+        if points < 0 and self.score < abs(points):
+            new_score = self.score = 0
+        else:
+            self.score += points
+            new_score = self.score
+
+        new_level = self.find_level_for_score(new_score)
+        if new_level.number > self.level.number:
+            self.level = new_level
+
+    def add_activity(self, points, activity_string):
+        """
+        Add activity to the newcomer.
+
+        This methof checks if the current activity is
+        already peformed by the user, if yes, then it
+        increase the 'number_of_times' field with one.
+        If not then it adds a new activity to the newcomer.
+        """
+        activity, created = Activity.objects.get_or_create(
+            name=activity_string, points=points)
+        if created:
+            activity.save()
+            self.activities.add(activity)
+        else:
+            activity.number_of_times += 1
+            activity.save()
+
+    def find_badges_for_activity(self, activities):
+        """
+        Find the badge based on the activities peformed by the newcomer.
+
+        :param activities: a QuerySet dict containing the 'name'
+                           as key and 'name of the activity' as value
+        :return: a badge object
+        """
+        activities = [activity['name'] for activity in activities]
+        badge = Badge.objects.filter(
+            b_activities__name__in=activities).distinct()
```

Comment:
I wonder if we could do it in one query, something like:
```
 badge = Badge.objects.filter(
            b_activities__all__name__in=activities).distinct()
```
Otherwise I am getting a long way:
```
    def find_badges_for_activity(self, activities):
        activities = [activity['name'] for activity in activities]
        badge_objects_list = []
        badges = Badge.objects.all()
        for badge in badges:
            b_activities = badge.b_activities.values('name')
            b_activities = [b_activity['name'] for b_activity in b_activities]
            match_activity_list = []
            for b_activity in b_activities:
                if b_activity in activities:
                    match_activity_list.append(1)
            if b_activities.count() == len(match_activity_list):
                badge_objects_list.append(badge)
       return badge_objects_list
```